src/Covenant/Test.hs

```diff
@@ -69,27 +126,435 @@ instance Arbitrary Concrete where
+            ]
+
+{-
+   Misc Helpers and the Arbitrary instances
+-}
+
+{- This saves us from having to write *another* newtype for each set of a "flavor" of datatypes and doesn't have any real drawbacks -}
+data DataDeclFlavor
+  = ConcreteDecl
+  | ConcreteNestedDecl
+  | SimpleRecursive
+  | Poly1
+
+-- can't be a Set b/c don't have the Ord instances, don't see what benefit a Vector gives since we'll almost always consume this
+-- via folding anyway
+newtype DataDeclSet (flavor :: DataDeclFlavor) = DataDeclSet [Covenant.Internal.Type.DataDeclaration AbstractTy]
+
+{- NOTE @Koz: This is supposed to be a "generic" shrinker for datatypes. It *should* return two paths:
+                - One that shrinks the number of constructors
+                - One that shrinks the constructors
+
+              This is why I had to add handling for `datatype` into `Concrete`. To use `shrink` recursively
+              on the structural components, we need some kind of instance to pivot off of. Since we want to avoid
+              writing a generic Arbitrary instance for Constructor or DataDeclaration, this seems like the
+              simplest solution.
+-}
+shrinkDataDecl :: Covenant.Internal.Type.DataDeclaration AbstractTy -> [Covenant.Internal.Type.DataDeclaration AbstractTy]
+shrinkDataDecl (Covenant.Internal.Type.DataDeclaration nm cnt ctors)
+  | Vector.null ctors = []
+  | otherwise =
+      let concreteShrink :: Vector (Covenant.Internal.Type.ValT AbstractTy) -> [Vector (Covenant.Internal.Type.ValT AbstractTy)]
+          concreteShrink = coerce . shrink . fmap Concrete
+
+          concreteShrink' :: Vector (Covenant.Internal.Type.ValT AbstractTy) -> [Vector (Covenant.Internal.Type.ValT AbstractTy)]
+          concreteShrink' xs = coerce $ fmap (Vector.fromList . shrink . Concrete) (Vector.toList xs)
+
+          ctorShrink :: Covenant.Internal.Type.Constructor AbstractTy -> [Covenant.Internal.Type.Constructor AbstractTy]
+          ctorShrink (Covenant.Internal.Type.Constructor ctorNm args) = Covenant.Internal.Type.Constructor ctorNm <$> concreteShrink args
+
+          ctorArgShrink :: Covenant.Internal.Type.Constructor AbstractTy -> [Covenant.Internal.Type.Constructor AbstractTy]
+          ctorArgShrink (Covenant.Internal.Type.Constructor ctorNm args) = Covenant.Internal.Type.Constructor ctorNm <$> concreteShrink' args
+
+          withShrunkenCtors = ((Covenant.Internal.Type.DataDeclaration nm cnt . Vector.fromList) . ctorShrink <$> Vector.toList ctors)
+
+          withShrunkenCtorArgs = Covenant.Internal.Type.DataDeclaration nm cnt . Vector.fromList . ctorArgShrink <$> Vector.toList ctors
+       in withShrunkenCtors <|> withShrunkenCtorArgs
+
+-- REVIEW: I dunno how liftShrink works under the hood so this might be redundant?
+shrinkDataDecls :: [Covenant.Internal.Type.DataDeclaration AbstractTy] -> [[Covenant.Internal.Type.DataDeclaration AbstractTy]]
+shrinkDataDecls decls = liftShrink shrinkDataDecl decls <|> (shrinkDataDecl <$> decls)
```

Comment:
`liftShrink` effectively provides the 'structural' shrink for a data type, and allows you to 'patch in' your own shrinker for 'contents'. So I'd say this is exactly the right thing here.

Reply:
Yay. I spent a while staring at QC source yesterday and was hoping that I got it right.